client-v2/src/main/java/com/clickhouse/client/api/Client.java

```diff
@@ -746,47 +746,113 @@ public CompletableFuture<InsertResponse> insert(String tableName,
-                future = request.data(stream.getInputStream()).execute();
-
-                //Copy the data from the input stream to the output stream
-                byte[] buffer = new byte[settings.getInputStreamCopyBufferSize()];
-                int bytesRead;
-                while ((bytesRead = data.read(buffer)) != -1) {
-                    stream.write(buffer, 0, bytesRead);
-                }
-            } catch (IOException e) {
-                responseFuture.completeExceptionally(new ClientException("Failed to write data to the output stream", e));
+            String retry = configuration.get(ClickHouseClientOption.RETRY.getKey());
+            final int maxRetries = retry == null ? (int) ClickHouseClientOption.RETRY.getDefaultValue() : Integer.parseInt(retry);
+            final int writeBufferSize = settings.getInputStreamCopyBufferSize() <= 0 ?
+                    Integer.parseInt(configuration.getOrDefault(ClickHouseClientOption.WRITE_BUFFER_SIZE.getKey(), "8192")) :
+                    settings.getInputStreamCopyBufferSize();
+
+            if (writeBufferSize <= 0) {
+                throw new IllegalArgumentException("Buffer size must be greater than 0");
             }
 
-            if (!responseFuture.isCompletedExceptionally()) {
-                try {
-                    int operationTimeout = getOperationTimeout();
-                    ClickHouseResponse clickHouseResponse;
-                    if (operationTimeout > 0) {
-                        clickHouseResponse = future.get(operationTimeout, TimeUnit.MILLISECONDS);
-                    } else {
-                        clickHouseResponse = future.get();
+            settings.setOption(ClickHouseClientOption.FORMAT.getKey(), format.name());
+            final InsertSettings finalSettings = settings;
+            CompletableFuture<InsertResponse> future = CompletableFuture.supplyAsync(() -> {
+                // Selecting some node
+                ClickHouseNode selectedNode = getNextAliveNode();
+
+                for (int i = 0; i <= maxRetries; i++) {
+                    // Execute request
+                    try (ClassicHttpResponse httpResponse =
+                                 httpClientHelper.insertRequest(selectedNode, finalSettings.getAllSettings(),
+                                         out -> {
+                                             out.write("INSERT INTO ".getBytes());
+                                             out.write(tableName.getBytes());
+                                             out.write(" \n FORMAT ".getBytes());
+                                             out.write(format.name().getBytes());
+                                             out.write(" \n".getBytes());
+
+                                             byte[] buffer = new byte[writeBufferSize];
+                                             int bytesRead;
+                                             while ((bytesRead = data.read(buffer)) != -1) {
```

Comment:
If we try to retry, do not we already exhaust the stream by doing `data.read`?

Reply:
Yes, this is correct. I need to try resetting the input stream. 
Will correct and create a test. 
However there is a story about retries and something smarter should be implemented
- check connection before use 
- cleanup connections by ttl